helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -294,99 +294,43 @@ private PartitionStateMap computeIntermediatePartitionState(ResourceControllerDa
       ClusterStatusMonitor clusterStatusMonitor, IdealState idealState, Resource resource,
       CurrentStateOutput currentStateOutput, PartitionStateMap bestPossiblePartitionStateMap,
       Map<String, List<String>> preferenceLists,
-      StateTransitionThrottleController throttleController) {
+      StateTransitionThrottleController throttleController, Map<Partition, List<Message>> resourceMessageMap) {
     String resourceName = resource.getResourceName();
     LogUtil.logDebug(logger, _eventId, String.format("Processing resource: %s", resourceName));
 
-    // Throttling is applied only on FULL-AUTO mode
-    if (!throttleController.isThrottleEnabled()
-        || !IdealState.RebalanceMode.FULL_AUTO.equals(idealState.getRebalanceMode())) {
+    // Throttling is applied only on FULL-AUTO mode and if the resource message map is empty, no throttling needed.
+    if (!throttleController.isThrottleEnabled() || !IdealState.RebalanceMode.FULL_AUTO.equals(
+        idealState.getRebalanceMode()) || resourceMessageMap.isEmpty()) {
       return bestPossiblePartitionStateMap;
     }
 
     String stateModelDefName = idealState.getStateModelDefRef();
     StateModelDefinition stateModelDef = cache.getStateModelDef(stateModelDefName);
-    PartitionStateMap intermediatePartitionStateMap = new PartitionStateMap(resourceName);
+    // This require a deep copy of current state map because some of the states will be overwritten by applying
+    // messages to it.
+    PartitionStateMap intermediatePartitionStateMap =
```

Comment:
Let me put a TODO here. If we do see this is a bottleneck, let's optimize it. This may require a complete design since in persistent stage, we need to output this result to ZK. And if you make it async, you cannot guarantee the correctness when you need to output in that stage.

Reply:
Sure, it can be future improvements. But I would suggest moving the intermediate state map calculation out of the compute() method. So it would be easier for use to make the change. Basically, compute() returns the message list. And we have another method to combine the messages to the current state and generate the intermediate state map in the execute() method.

Regarding my proposal, I searched the related code:
There are 3 stages need the intermediate state map, all of them are async. So I think it would be safe to do so. At least it is the current assumption that the check or persistence is done asynchronously.

I'm not 100% sure about the monitoring part. My quick search shows that all the related monitoring items do not require you to compute the full intermediate state map. The messages would be enough.

Moreover, regarding the 3 stages,
1. PersistAssignmentStage needs the whole map, so I suggest we do the intermediate state calculation there.
2. TargetExteralViewCalcStage also needs the whole map, I'm not sure if we should do the calculation 2 times to avoid latency or adding a fancier design so that only one of them finishes the calculation, and then we are good.
3. MaintenanceRecoveryStage needs the map to check MAX_PARTITION_PER_INSTANCE_EXCEEDED violation, but I'm pretty sure it can also be done by checking the current state + the message list. In which case, you don't need to copy-paste the state map and process it.

Reply:
Moved the derive logic to another function.